File: Leetcode/397-Integer-Replacement/IntegerReplacement.js
/**
 * @param {number} n
 * @return {number}
 */
var integerReplacement = function(n) {
    let dp = []

    function solve(num) {
        if(num==1) {
            return 0
        }
        if(dp[num]) {
            return dp[num]
        }
        if(num%2==0) {
            dp[num] = solve(num/2)+1
        } else {
            // odd number
            let a = solve(num-1) +1
            let b = solve((num+1)/2)+2
            dp[num] = Math.min(a,b)
        }
        return dp[num]
    }

    return solve(n)
};

// TLE
// var integerReplacement = function(n) {
//     var count = 0
//     while (n!=1) {
//         var small = n&(n-1)
//         var big = n&(n+1)
//         if ((n&1)==0) {
//             n>>=1
//         } else if ((small&(small-1)) <= (big&(big-1))) {
//             n-=1
//         } else {
//             n+=1
//         }
//         count+=1
//     }
//     return count
// };

File: Leetcode/397-Integer-Replacement/IntegerReplacement.cpp
class Solution {
public:
    int integerReplacement(int n) {
        long int N = n;
        long int small, big;
        int count = 0;
        while (N!=1) {
            small = N & (N-1);
            big = N & (N+1);
            if ((N&1) ==0) {
                N >>=1;
            } else if ( (small&(small-1))<=(big&(big-1)) ) {
                N -=1;
            } else {
                N +=1;
            }
            count++;
        }
        return count;
    }
};
